src/linalg.h

```diff
@@ -75,7 +80,27 @@ inline void matrix_multiplication(T** matrix, T* vector_x,
 
 
 /**
- * @brief Scale vectgor by some scalar value.
+ * @brief Transpose a 2D matrix.
+ * @param matrix array to transpose
+ * @param dim1 first dimension length
+ * @param dim2 second dimension length
+ */
+template<typename T>
+inline void matrix_transpose(T* matrix, int dim1, int dim2) {
+
+  std::vector<T> temp;
+
+  for (int i=0; i < dim1; i++) {
+    for (int j=0; j < dim2; j++)
+      temp.push_back(matrix[j * dim1 + i]);
+  }
```

Comment:
Excellent point, the ``temp`` vector now calls ``resize()`` to allocate the appropriate amount of memory upon initialization. I agree that this is almost guaranteed to not be the most cache efficient implementation of a matrix transpose. But since this is definitely not a performance critical section of code I just went with the most intuitive implementation that came to mind.